src/Common/src/System/Diagnostics/Debug.cs

```diff
@@ -84,19 +144,33 @@ public static void WriteLine(string message)
         [System.Diagnostics.Conditional("DEBUG")]
         public static void Write(string message)
         {
-            s_logger.WriteCore(message ?? string.Empty);
+            if (message == null)
+            {
+                WriteCore(string.Empty);
+                return;
+            }
+            if (s_needIndent)
+            {
+                message = GetIndentString() + message;
+                s_needIndent = false;
+            }
+            WriteCore(message);
+            if (message.EndsWith(Environment.NewLine))
+            {
+                s_needIndent = true;
```

Comment:
> The implementation in DefaultTraceListener also doesn't synchronize at all.

The implementation of Debug does, though, via TraceInternal, e.g. look at all the locking done in https://referencesource.microsoft.com/#System/compmod/system/diagnostics/TraceInternal.cs,539a95585eee5404

Further, IndentLevel is stored in a [ThreadStatic]:
https://referencesource.microsoft.com/#System/compmod/system/diagnostics/TraceInternal.cs,539a95585eee5404